internal/api/ratelimits.go

```diff
@@ -0,0 +1,27 @@
+package api
+
+import (
+	"github.com/supabase/auth/internal/conf"
+	"golang.org/x/time/rate"
+)
+
+// newRateLimiter returns a rate limiter configured using the given conf.Rate.
+//
+// The returned *rate.Limiter will be configured with a token bucket containing
+// a single token, which will fill up at a rate of r. For example to allow 100
+// events every 24 hours. This will fill a token bucket approximately once every
+// 864 seconds (14.4 minutes). See Example_newRateLimiter for a visualization.
+func newRateLimiter(r conf.Rate) *rate.Limiter {
+	// The rate limiter deals in events per second.
+	eps := r.EventsPerSecond()
+	burst := int(r.Events)
+	if burst <= 0 {
+		burst = 1
+	}
+
+	// NewLimiter will have an initial token bucket of size `burst`. It will
+	// be refilled at a rate of `eps` indefinitely. Note that the expression
+	// 100 / 24h is roughly equivelant to the expression 1 / 15m. The 100 is
+	// a rate, not a quota.
+	return rate.NewLimiter(rate.Limit(eps), burst)
```

Comment:
This is exactly what we don't want with this change.

If the setting is `100/24h` we need a token bucket that refills with 100 tokens every 24 hours. Not a token bucket that fills up `100 / (24 * 60 * 6) = 0.0011574074074074073` tokens per second.

This means that:

- The server can use up 100 events instantly or
- 99 events in the first second and 1 after `24h - 1s`
- 1 event every 15 minutes

Unfortunately the `rate` package does not do this. It very strictly operates with a token bucket with a fixed refill rate of 1s. We need a refill rate that is controlled by the second part of the config `<events> / <refill-rate>`. I don't think this is possible with this package.

Let's go back to the original idea by implementing our own variable-duration rate limiter, because:

- It's simpler because it's much easier to reason about:
  - Atomic counter, no need for complex synchronization logic
  - The counter resets at the specified interval by way of a simple goroutine or maybe using the [rate.Sometimes](https://pkg.go.dev/golang.org/x/time/rate#Sometimes) object
- Works as intended in real life, not mathematically
  - Here is a demonstration that does not simulate the behavior we want: https://gist.github.com/hf/16fae8179baa65dde76b6147a356af75
- We won't be willing the rate.Limiter object into submission
- Very easy to write tests for, unlike with the rate.Limiter object
- Incredibly easy to explain to customers how to customize this for their needs

Reply:
I've implemented it exactly as requested. Let me know if you want me to change it to start a goroutine and fill it instead.